ge/org.osate.ge/src/org/osate/ge/internal/ui/tools/ToolUtil.java

```diff
@@ -80,4 +108,234 @@ public static Context findContextExcludeOwner(final BusinessObjectContext boc,
+		addToBuilder(diagnostics, syntaxDiagnostics);
+
+		final EObject serializedObject = resourceSet.getEObject(EcoreUtil.getURI(eObjectToValidate), true);
+		final Diagnostic validationDiagnostic = Diagnostician.INSTANCE.validate(serializedObject,
+				Collections.singletonMap(Diagnostician.VALIDATE_RECURSIVELY, true));
+		if (isErrorOrWarning(validationDiagnostic.getSeverity())) {
+			// Validation Diagnostics
+			final Stream<DiagnosticWrapper> qualifiedDiagnostics = getDiagnosticDescendants(validationDiagnostic)
+					.filter(diagnostic -> diagnostic instanceof FeatureBasedDiagnostic)
+					.map(diagnostic -> diagnosticBuilder.build(diagnostic));
+			addToBuilder(diagnostics, qualifiedDiagnostics);
+		}
+
+		// Errors
+		final Stream<DiagnosticWrapper> resourceErrors = getResourceDiagnostics(resource.getErrors(),
+				diagnosticBuilder,
+				Diagnostic.ERROR);
+		addToBuilder(diagnostics, resourceErrors);
+
+		// Warnings
+		final Stream<DiagnosticWrapper> resourceWarnings = getResourceDiagnostics(resource.getWarnings(),
+				diagnosticBuilder,
+				Diagnostic.WARNING);
+		addToBuilder(diagnostics, resourceWarnings);
+
+		return diagnostics.build().collect(Collectors.toSet());
+	}
+
+	private static void addToBuilder(final Builder<DiagnosticWrapper> builder,
+			final Stream<DiagnosticWrapper> diagnosticsToAdd) {
+		diagnosticsToAdd.forEach(diagnostic -> builder.add(diagnostic));
+	}
+
+	private static class DiagnosticBuilder {
+		private final String msgPrefix;
+
+		public DiagnosticBuilder(final String msgPrefix) {
+			this.msgPrefix = msgPrefix;
+		}
+
+		public DiagnosticWrapper build(final int severity, final String msg) {
+			return new DiagnosticWrapper(severity, getDetailedDiagnosticMessage(this.msgPrefix, msg));
+		}
+
+		public DiagnosticWrapper build(final Diagnostic diagnostic) {
+			return new DiagnosticWrapper(diagnostic.getSeverity(),
+					getDetailedDiagnosticMessage(this.msgPrefix, diagnostic.getMessage()));
+		}
+
+		public DiagnosticWrapper build(final int severity,
+				final org.eclipse.emf.ecore.resource.Resource.Diagnostic diagnostic) {
+			return new DiagnosticWrapper(severity,
+					getDetailedDiagnosticMessage(this.msgPrefix, diagnostic.getMessage()));
+		}
+
+		private static String getDetailedDiagnosticMessage(final String prefix, final String message) {
+			return new StringBuilder(prefix).append(message).toString();
+		}
+	}
+
+	private static class DiagnosticWrapper extends BasicDiagnostic {
+		public DiagnosticWrapper(final int severity, final String message) {
+			this.severity = severity;
+			this.message = message;
+		}
+
+		@Override
+		public boolean equals(final Object o) {
+			if (o == null) {
+				return false;
+			}
+			if (o == this) {
+				return true;
+			}
+			if (getClass() != o.getClass()) {
+				return false;
+			}
+
+			final DiagnosticWrapper e = (DiagnosticWrapper) o;
+			return (this.message.equalsIgnoreCase(e.getMessage()));
+		}
+
+		@Override
+		public int hashCode() {
```

Comment:
Why are hashCode and equals being overriden? This implementation violates the contract with `hashCode` Equal object must return the same hashCode.

Reply:
One way to make it consistent is to store a lowercase version of the message as  a private field and to use it in hashcode and equals.

Reply:
I made changes based on the first comment, probably shouldn't be ignoring severity anyways.

Reply:
Well, I don't think my original comment was clear. The original problem wasn't the severity, it was that hashCode was that equals was doing a case insensitive comparison and that wasn't consistent with hashCode. Although I agree that severity shouldn't be ignored.

Reply:
Yeah, I know, my updated code fixes both issues right?

Reply:
Correct but I wanted to be clear on the original issue.

That said, I'm not sure if all the changes are beneficial. The new name `BasicDiagnostic` is confusing since EMF already has one and the new javadoc isn't very helpful. All diagnostics contain severity and a message. Based on your earlier comments, the purpose of this diagnostic implementation is to avoid duplicates. That's the important part and it is not captured in the comments.

However, I think there may be a better way of handling things. I realize this may require changes in several places, but what do you think about using TreeSet with a custom comparator instead of HashSet. The comparator would compare the diagnostics in whatever way is appropriate. Just comparing may be okay.. If you did this, you wouldn't need the new diagnostic type. The builder could just create EMF BasicDiagnostic directly.  Thoughts? 

In any case, let me know when you finish the user guide updates and I'll do a final review.

Reply:
Well, XtextResource returns a org.eclipse.emf.ecore.resource.Resource.Diagnostic

Reply:
I'm not sure if I see the issue.

Diagnostic isn't a class, it is an interface. Ideally we would write everything to use Diagnostic. You are adding a prefix so you subclassed EMF BasicDiagnostic. Correct? However, if you use TreeSet with a comparator, you can continue to use EMF's `BasicDiagnostic` without our separate class.  Whatever route you take, we need a comment/javadoc explaining why we are doing it that way.